src/OFS/CopySupport.py

```diff
@@ -587,7 +587,35 @@ def _getCopy(self, container):
         f.seek(0)
         ob=container._p_jar.importFile(f)
         f.close()
-        return ob
+        # Cleanup the copy.  It may contain private objects that the current
+        # user is not allowed to see.
+        sm = getSecurityManager()
+        if not sm.checkPermission('View', self):
+            # The user is not allowed to view the object that is currently
+            # being copied, so it makes no sense to check any of its sub
+            # objects.  It probably means we are in a test.
+            return ob
+        return self._cleanupCopy(ob)
+
+    def _cleanupCopy(self, cp):
+        sm = getSecurityManager()
+        ob = aq_base(self)
+        if hasattr(ob, 'objectIds'):
+            for k in self.objectIds():
+                v = self._getOb(k)
+                if not sm.checkPermission('View', v):
+                    # We do not use cp._delObject, because this would fire
+                    # events that are needless for objects that are not even in
+                    # an Acquisition chain yet.
+                    cp._delOb(k)
```

Comment:
ISTM we should be logging something about the skipped / deleted objects, so that the user has at least some chance to figure out what happened to their not-quite-a-copy.